### PR TITLE
azurerm_spring_cloud_service supports connection_string

### PR DESCRIPTION
**Pull request by ms-henglu**
image

### CODE REVIEW

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 1]
```diff
@@ -202,8 +203,18 @@ func resourceSpringCloudService() *pluginsdk.Resource {
 				Elem: &pluginsdk.Resource{
 					Schema: map[string]*pluginsdk.Schema{
 						"instrumentation_key": {
-							Type:     pluginsdk.TypeString,
-							Required: true,
+							Type:         pluginsdk.TypeString,
+							Optional:     true,
+							Deprecated:   "Deprecated in favour of `connection_string`",
```
Comment (katbyte) [inline internal/services/springcloud/spring_cloud_service_resource.go:208]: could we explain why and how to migrate?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @ms-henglu - could we explain why we are making this change here?

**Comment by ms-henglu** [comment]
@katbyte , Thanks!

I've added how to migrate, but I'm not sure how to explain why. It's a server side change, it works on both connection string and instrumentation key, but it's recommended to use connection string.

**Comment by ms-henglu** [comment]
> But why is it recommended and we are forcing users to change? what is the difference and given both exist for the app insites resource should we not support both? or should it be deprecated there too?
> 
> ```
>     instrumentation_key = azurerm_application_insights.test.instrumentation_key
>     connection_string = azurerm_application_insights.test.connection_string
> ```

It's server side change on Azure Spring Cloud. It will let user to pass app insights's connection_string to spring cloud when it gets GA. But to keep backward combability, it still supports instrumentation key.

On the side of app insights, nothing changed, we don't need to modify app insights.

**Comment by katbyte** [comment]
At some point in the future will conntection_string stop accepting the instrumentation key?

**Comment by ms-henglu** [comment]
> At some point in the future will conntection_string stop accepting the instrumentation key?

Yes, but not in near feature.

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 2]
```diff
@@ -202,8 +203,18 @@ func resourceSpringCloudService() *pluginsdk.Resource {
 				Elem: &pluginsdk.Resource{
 					Schema: map[string]*pluginsdk.Schema{
 						"instrumentation_key": {
-							Type:     pluginsdk.TypeString,
-							Required: true,
+							Type:         pluginsdk.TypeString,
+							Optional:     true,
+							Deprecated:   "Deprecated in favour of `connection_string` which is the connection string for Application Insights",
```
Comment (katbyte) [inline internal/services/springcloud/spring_cloud_service_resource.go:208]: lets be clear here then
```suggestion
							Deprecated:   "This property is due to be removed from the service API and thus hs been deprecated. Please switch to using the `connection_string` property with the connection string for the Application Insights instance.",
```

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 3]
```diff
@@ -651,9 +662,16 @@ func expandSpringCloudTrace(input []interface{}) *appplatform.MonitoringSettingP
...
+	if value := v["instrumentation_key"]; len(value.(string)) > 0 {
+		instrumentationKey = value.(string)
+	}
+	if value := v["connection_string"]; len(value.(string)) > 0 {
+		instrumentationKey = value.(string)
+	}
 	return &appplatform.MonitoringSettingProperties{
 		TraceEnabled:                  utils.Bool(true),
-		AppInsightsInstrumentationKey: utils.String(v["instrumentation_key"].(string)),
+		AppInsightsInstrumentationKey: utils.String(instrumentationKey),
```
Comment (katbyte) [inline internal/services/springcloud/spring_cloud_service_resource.go:674]: is there no new property on the backend for this/is it forthcoming in a new API version?

**Review by katbyte — CHANGES_REQUESTED** [review]

File: internal/services/springcloud/spring_cloud_service_resource.go (hunk 3 above)
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_service_resource.go:674]: No, there isn't.

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 4]
```diff
@@ -202,8 +203,18 @@ func resourceSpringCloudService() *pluginsdk.Resource {
 				Elem: &pluginsdk.Resource{
 					Schema: map[string]*pluginsdk.Schema{
 						"instrumentation_key": {
-							Type:     pluginsdk.TypeString,
-							Required: true,
+							Type:         pluginsdk.TypeString,
+							Optional:     true,
+							Deprecated:   "This property is due to be removed from the service API and thus hs been deprecated. Please switch to using the `connection_string` property with the connection string for the Application Insights instance.",
```
Comment (katbyte) [inline internal/services/springcloud/spring_cloud_service_resource.go:208]:
```suggestion
							Deprecated:   "This property is due to be removed from this service's API and thus has been deprecated and will be removed in v3.0 of the provider. Please switch to using the `connection_string` property with the connection string for the Application Insights instance to use.",
```

**Review by katbyte — APPROVED** [review]
Aside from one comment LGTM 🏗️

**Pull request merged**